### PR TITLE
ARROW-12012: [Java][JDBC] Fix BinaryConsumer reallocation

### PR DESCRIPTION
**Pull request by zxf**
[ARROW-12012](https://issues.apache.org/jira/browse/ARROW-12012)
An exception will be thrown when BinaryConsumer consumes a large amount or a lot of data.

### CODE REVIEW

**Comment by emkornfield** [comment]
@liyafan82 or @tianchen92 do you have time to review?

**Comment by liyafan82** [comment]
> @liyafan82 or @tianchen92 do you have time to review?

@zxf Thanks for the patch. I will take a look.

File: java/adapter/jdbc/src/main/java/org/apache/arrow/adapter/jdbc/consumer/BinaryConsumer.java [hunk 1]
```diff
@@ -57,30 +55,44 @@ public BinaryConsumer(VarBinaryVector vector, int index) {
...
+  public static void setSafe(VarBinaryVector vector, int index, InputStream is) throws IOException {
+    assert index >= 0;
+    while (index >= vector.getValueCapacity()) {
+      vector.reallocValidityAndOffsetBuffers();
+    }
+    final int startOffset = vector.getStartOffset(index);
+    final ArrowBuf offsetBuffer = vector.getOffsetBuffer();
+    int dataLength = 0;
+    int read;
+    byte[] bytes = new byte[BUFFER_SIZE];
```
Comment (liyafan82) [inline java/adapter/jdbc/src/main/java/org/apache/arrow/adapter/jdbc/consumer/BinaryConsumer.java:77]: Repeated small allocations can be expensive. Can we reuse the buffer?

File: java/adapter/jdbc/src/main/java/org/apache/arrow/adapter/jdbc/consumer/BinaryConsumer.java [hunk 2]
```diff
@@ -57,30 +55,44 @@ public BinaryConsumer(VarBinaryVector vector, int index) {
...
-          vector.reallocDataBuffer();
-        }
-        PlatformDependent.copyMemory(bytes, 0,
-                dataBuffer.memoryAddress() + startIndex + totalBytes, read);
-        totalBytes += read;
-      }
-      offsetBuffer.setInt((currentIndex + 1) * 4, startIndex + totalBytes);
-      BitVectorHelper.setBit(vector.getValidityBuffer(), currentIndex);
-      vector.setLastSet(currentIndex);
+      setSafe(vector, currentIndex, is);
```
Comment (liyafan82) [inline java/adapter/jdbc/src/main/java/org/apache/arrow/adapter/jdbc/consumer/BinaryConsumer.java:95]: nit: it seems this method is simply enough. why extract another method?

File: java/adapter/jdbc/src/test/java/org/apache/arrow/adapter/jdbc/consumer/BinaryConsumerTest.java [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.arrow.adapter.jdbc.consumer;
+
+import static org.junit.Assert.*;
```
Comment (liyafan82) [inline java/adapter/jdbc/src/test/java/org/apache/arrow/adapter/jdbc/consumer/BinaryConsumerTest.java:20]: Please split this into a number of import statements for each class.

File: java/adapter/jdbc/src/test/java/org/apache/arrow/adapter/jdbc/consumer/BinaryConsumerTest.java [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+public class BinaryConsumerTest extends AbstractConsumerTest {
+
+  private static final int INITIAL_VALUE_ALLOCATION = BaseValueVector.INITIAL_VALUE_ALLOCATION;
+  private static final int DEFAULT_RECORD_BYTE_COUNT = 8;
+
+  protected void assertConsume(boolean nullable, Consumer<BinaryConsumer> dataConsumer, byte[][] expect) {
+    try (final VarBinaryVector vector = new VarBinaryVector("binary", allocator)) {
+      BinaryConsumer consumer = BinaryConsumer.createConsumer(vector, 0, nullable);
+      dataConsumer.accept(consumer);
+      assertEquals(expect.length - 1, vector.getLastSet());
```
Comment (liyafan82) [inline java/adapter/jdbc/src/test/java/org/apache/arrow/adapter/jdbc/consumer/BinaryConsumerTest.java:39]: we also need `assertEquals(expect.length, vector.getValueCount);` ?

File: java/adapter/jdbc/src/test/java/org/apache/arrow/adapter/jdbc/consumer/BinaryConsumerTest.java [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+  }
+
+
+  public void testConsumeInputStream(byte[][] values, boolean nullable) throws IOException {
+    assertConsume(nullable, binaryConsumer -> {
+      for (byte[] value : values) {
+        try {
+          binaryConsumer.consume(new ByteArrayInputStream(value));
+        } catch (IOException e) {
+          e.printStackTrace();
```
Comment (liyafan82) [inline java/adapter/jdbc/src/test/java/org/apache/arrow/adapter/jdbc/consumer/BinaryConsumerTest.java:66]: why swallow the exception?

File: java/adapter/jdbc/src/test/java/org/apache/arrow/adapter/jdbc/consumer/BinaryConsumerTest.java [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+        createBytes(DEFAULT_RECORD_BYTE_COUNT)
+    }, false);
+
+    testConsumeInputStream(new byte[][]{
+        createBytes(INITIAL_VALUE_ALLOCATION * DEFAULT_RECORD_BYTE_COUNT)
+    }, false);
+
+    testConsumeInputStream(new byte[][]{
+        createBytes(INITIAL_VALUE_ALLOCATION * DEFAULT_RECORD_BYTE_COUNT * 10),
+    }, false);
```
Comment (liyafan82) [inline java/adapter/jdbc/src/test/java/org/apache/arrow/adapter/jdbc/consumer/BinaryConsumerTest.java:96]: do we need a case for `nullable = true`

Comment (zxf) [reply java/adapter/jdbc/src/test/java/org/apache/arrow/adapter/jdbc/consumer/BinaryConsumerTest.java:97]: According to the original logic, when a nullable consumer processes a null value, it just adds 1 to its own index and does not modify the content of the vector. And set the rowCount in ArrowVectorIterator.

File: java/adapter/jdbc/src/test/java/org/apache/arrow/adapter/jdbc/consumer/BinaryConsumerTest.java (hunk 4 above)
Comment (zxf) [reply java/adapter/jdbc/src/test/java/org/apache/arrow/adapter/jdbc/consumer/BinaryConsumerTest.java:44]: The valueCount is set outside the consumer. So if only the consumer is called, the valueCount will always be zero.

**Comment by liyafan82** [comment]
Rerunning integration tests. The failure seems unrelated.

**Comment by liyafan82** [comment]
Mostly look good. Thank you @zxf

**Comment by liyafan82** [comment]
+1. Thanks @zxf

**Pull request closed without merging**